backend/service/service_tx.go

```diff
@@ -316,6 +316,92 @@ func (service *TxService) QueryTxNumGroupByDay() []model.TxNumGroupByDayVo {
 	return result
 }
 
+func (service *TxService) ParseCoinFlowFromAndTo(txType, from, to string) (string, string) {
+	switch txType {
+	case types.TxTypeTransfer:
+		return from, to
+	case types.TxTypeBurn:
+		return from, ""
+	case types.TxTypeStakeEditValidator:
+		return "", ""
+	case types.TxTypeStakeDelegate:
+		return from, to
+	case types.TxTypeUnjail:
+		return "", ""
+	case types.TxTypeSetWithdrawAddress:
+		return "", ""
+	case types.TxTypeStakeCreateValidator:
+		return from, to
+		//exchange
+	case types.TxTypeBeginRedelegate:
+		return to, from
+	case types.TxTypeWithdrawDelegatorReward:
+		return to, ""
+	case types.TxTypeWithdrawDelegatorRewardsAll:
+		return "", ""
+	case types.TxTypeWithdrawValidatorRewardsAll:
+		return "", ""
+	case types.TxTypeStakeBeginUnbonding:
+		return to, from
+	default:
+		return from, to
+	}
+}
+
+func (service *TxService) QueryTxMsgByHashArr(hashArr []string) []document.TxMsg {
+
+	selector := bson.M{
+		document.TxMsg_Field_Hash:    1,
+		document.TxMsg_Field_Content: 1,
+		document.TxMsg_Field_Type:    1,
+	}
+	condition := bson.M{
+		document.TxMsg_Field_Hash: bson.M{"$in": hashArr},
+	}
+
+	txMsgs := []document.TxMsg{}
+	if err := queryAll(document.CollectionNmTxMsg, selector, condition, "", 0, &txMsgs); err != nil {
+		logger.Error("query tx msg", logger.String("err", err.Error()))
+		panic(types.CodeNotFound)
+	}
+
+	return txMsgs
+}
+
+func (service *TxService) getForwardAddr(txType, content string) (string, error) {
+	m := make(map[string]interface{})
+	err := json.Unmarshal([]byte(content), &m)
+	if err != nil {
+		return "", err
+	}
+
+	switch txType {
+	case types.TxTypeBeginRedelegate:
+		if v, ok := m["validator_src_addr"].(string); ok {
+			return v, nil
+		}
+	case types.TxTypeWithdrawDelegatorReward:
```

Comment:
there is no need to get fva addr from tx msg when tx type is WithdrawDelegatorReward and WithdrawValidatorRewardsAll, you can get f/iva addr from tx_common